subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/PropertyValidationAccess.java

```diff
@@ -62,62 +62,35 @@ public static void collectTaskValidationProblems(Class<?> topLevelBean, Map<Stri
             new ClasspathPropertyAnnotationHandler(), new CompileClasspathPropertyAnnotationHandler()
         ));
         Queue<BeanTypeNode> queue = new ArrayDeque<BeanTypeNode>();
-        queue.add(BeanTypeNode.create(null, TypeToken.of(topLevelBean)));
+        queue.add(BeanTypeNode.create(null, TypeToken.of(topLevelBean), metadataStore));
         boolean cacheable = taskClassInfoStore.getTaskClassInfo(Cast.<Class<? extends Task>>uncheckedCast(topLevelBean)).isCacheable();
 
         while (!queue.isEmpty()) {
             BeanTypeNode node = queue.remove();
-            validateTaskClass(topLevelBean, cacheable, problems, node, new NestedTypeContext(metadataStore, queue), metadataStore.getTypeMetadata(node.getBeanClass()));
+            node.visit(topLevelBean, cacheable, problems, queue, metadataStore);
         }
     }
 
-    private static void validateTaskClass(Class<?> beanClass, boolean cacheable, Map<String, Boolean> problems, BeanTypeNode node, NestedPropertyContext<BeanTypeNode> context, TypeMetadata typeMetadata) {
-        for (PropertyMetadata metadata : typeMetadata.getPropertiesMetadata()) {
-            String qualifiedPropertyName = node.getQualifiedPropertyName(metadata.getFieldName());
-            for (String validationMessage : metadata.getValidationMessages()) {
-                problems.put(propertyValidationMessage(beanClass, qualifiedPropertyName, validationMessage), Boolean.FALSE);
-            }
-            Class<? extends Annotation> propertyType = metadata.getPropertyType();
-            if (propertyType == null) {
-                if (!Modifier.isPrivate(metadata.getMethod().getModifiers())) {
-                    problems.put(propertyValidationMessage(beanClass, qualifiedPropertyName, "is not annotated with an input or output annotation"), Boolean.FALSE);
+    private abstract static class BeanTypeNode extends AbstractPropertyNode {
+
+        public static BeanTypeNode create(@Nullable String parentPropertyName, TypeToken<?> beanType, PropertyMetadataStore metadataStore) {
+            TypeMetadata typeMetadata = metadataStore.getTypeMetadata(beanType.getRawType());
```

Comment:
`beanType.getRawType()` is not free, it actually does a lookup, and creates an `Iterator`. It should be extracted to a local variable.

Reply:
Extracted.